extensions/amp-carousel/0.1/slidescroll.js

```diff
@@ -927,4 +1019,104 @@ export class AmpSlideScroll extends BaseSlides {
   analyticsEvent_(eventType, vars) {
     triggerAnalyticsEvent(this.element, eventType, vars);
   }
+
+  /**
+   * Sets up the `autoplay` configuration.
+   * @private
+   */
+  setupAutoplay_() {
+    const delayValue = Number(this.element.getAttribute('delay'));
+    // If it isn't a number and is not greater than 0 then don't assign
+    // and use the default.
+    if (delayValue > 0) {
+      // Guard against autoplayValue that is lower than 1s to prevent
+      // people from crashing the runtime with providing very low delays.
+      this.autoplayDelay_ = Math.max(1000, delayValue);
+    }
+
+    // By default `autoplay` should also mean that the current carousel slide
+    // is looping. (to be able to advance past the last item)
+    if (!this.hasLoop_) {
+      this.element.setAttribute('loop', '');
+      this.loopAdded_ = true;
+      this.hasLoop_ = true;
+      this.shouldLoop = true;
+    }
+  }
+
+  /**
+   * Starts the autoplay delay if allowed.
+   * @private
+   */
+  autoplay_() {
+    if (!this.shouldAutoplay_ || this.autoplayLoops_ == 0) {
+      return;
+    }
+    this.clearAutoplay_();
+    this.autoplayTimeoutId_ = /** @type {number} */ (
+      Services.timerFor(this.win).delay(
+        this.go.bind(
+          this,
+          /* dir */ 1,
+          /* animate */ true,
+          /* autoplay */ true
+        ),
+        this.autoplayDelay_
+      )
+    );
+  }
+
+  /**
+   * Called by toggleAutoplay action to toggle the autoplay feature.
+   * @param {boolean} toggleOn
+   * @private
+   */
+  toggleAutoplay_(toggleOn) {
+    if (toggleOn == this.shouldAutoplay_) {
+      return;
+    }
+
+    const prevAutoplayStatus = this.shouldAutoplay_;
+
+    this.hasAutoplay_ = toggleOn;
+    this.shouldAutoplay_ = this.hasAutoplay_ && this.isLoopingEligible();
+
+    if (!prevAutoplayStatus && this.shouldAutoplay_) {
+      this.setupAutoplay_();
+    }
+
+    if (this.shouldAutoplay_) {
+      this.autoplay_();
+    } else {
+      this.clearAutoplay_();
+    }
+  }
+
+  /**
+   * Clear the autoplay timer.
+   * @private
+   */
+  clearAutoplay_() {
```

Comment:
nit: it's not obvious the difference between `clearAutoplay` and `removeAutoplay`, can we rename?
```suggestion
  clearAutoplayTimer_() {
```

Reply:
It still isn't obvious to me....why are the timer and the attributes handled in separate funcs?

Reply:
I'm not sure either. `clearAutoplay` appears to be called much more frequently than `removeAutoplay`...